### DISCUSSION TITLE
Mirth XML CCD troubles

### DISCUSSION THREAD

**Discussion by jeffkirkley** [nextgenhealthcare/connect#5888] [Q&A]
Hello everyone.  I posted this on the Mirth forums site, but not sure if it is getting a lot of visibility.

***Disclaimer:  I am VERY new to Mirth and HL7 and CCD.  I have a general understanding, but this is my very first exposure to all of this.

Below is what I am trying to do:

I have the partial XML CCD snippet below:

```xml
 
<xml>
<document>
<ClinicalDocument xmlns="urn:hl7-org:v3" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" moodCode="EVN">
<realmCode code="US"/>
<typeId extension="POCD_HD000040" root="2.16.840.1.113883.1.3"/>
<templateId root="2.16.840.1.113883.10.20.22.1.1"/>
<templateId root="2.16.840.1.113883.10.20.22.1.2"/>
<id extension="XXXXXXXXXXXXXXXXXXX" root="2.16.840.1.113883.3.929.4.158610.100"/>
<code code="34133-9" codeSystem="2.16.840.1.113883.6.1"/>
<title>Clinical Document</title>
<effectiveTime value="20230810134929"/>
<confidentialityCode code="N" codeSystem="2.16.840.1.113883.5.25"/>
<languageCode code="en-US"/>
<recordTarget contextControlCode="OP" typeCode="RCT">
<patientRole classCode="PAT">
<id assigningauthority="XXXXX" extension="0123456789" root="2.16.840.1.113883.4.290"/>
<addr>
```

I am trying to write in javascript in a Mirth destination transformer a code section to produce something similar to:

```xml
<xml>
<document>
<ClinicalDocument xmlns="urn:hl7-org:v3" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" moodCode="EVN">
<realmCode code="US"/>
<typeId extension="POCD_HD000040" root="2.16.840.1.113883.1.3"/>
<templateId root="2.16.840.1.113883.10.20.22.1.1"/>
<templateId root="2.16.840.1.113883.10.20.22.1.2"/>
<id extension="XXXXXXXXXXXXXXXXXXX" root="2.16.840.1.113883.3.929.4.158610.100"/>
<code code="34133-9" codeSystem="2.16.840.1.113883.6.1"/>
<title>Clinical Document</title>
<effectiveTime value="20230810134929"/>
<confidentialityCode code="N" codeSystem="2.16.840.1.113883.5.25"/>
<languageCode code="en-US"/>
<recordTarget contextControlCode="OP" typeCode="RCT">
<patientRole classCode="PAT">
<id assigningauthority="XXXXX" extension="0123456789" root="2.16.840.1.113883.4.290"/> //shows original info
<id assigningauthority="YYYYY" extension="9876543210" root="2.16.840.1.113883.4.290"/> //shows modified info
<addr>
```

I have looked all over the boards here and on Google and have tried the various insertchildbefore, insertchildafter, and append routines, but nothing seems to want to play nice with the XML formatting/structure. For above, need both lines present in transformed output.

This is the javascript code I am running in a destination transformer:

```
var receiveMrn = msg['receivemrn'].toString()
channelMap.put('receiveMrn', receiveMrn);

var patientId = $('receiveMrn');
channelMap.put('patientId', patientId);

var assigningAuthority = msg['assigningauthority'].toString()
channelMap.put('assigningAuthority', assigningAuthority);

msg['document'].*::['ClinicalDocument'].*::['recordTarget'].*::['patientRole'].*::['id']['@extension'] = $('patientId');

msg['document'].*::['ClinicalDocument'].*::['recordTarget'].*::['patientRole'].*::['id']['@root'] = '2.16.840.1.113883.4.290';

msg['document'].*::['ClinicalDocument'].*::['recordTarget'].*::['patientRole'].*::['id']['@assigningauthority'] = $('assigningAuthority');

```

After a bunch of trial and error, I came up with the following:

```xml
<languageCode code="en-US"/>
<recordTarget contextControlCode="OP" typeCode="RCT">
<patientRole classCode="PAT">
<id assigningauthority="XXXXX" extension="0123456789" root="2.16.840.1.113883.4.290">
<id2 xmlns="" extension="2023-08-15" root="2.16.840.1.113883.10.20.22.1.1"/>
</id>
<addr>
```

```
var addTemplateId = new XML('<id2/>');

addTemplateId['@root'] = '2.16.840.1.113883.10.20.22.1.1';
addTemplateId['@extension'] = '2023-08-15';

var newValue = msg['document'].*::['ClinicalDocument'].*::['recordTarget'].*::['patientRole'].*::['id'].appendChild(addTemplateId);
```
​

My question now is how can I remove:

xmlns=""

**Comment by pacmano1** [comment]
Relevant forum link: https://forums.mirthproject.io/forum/mirth-connect/support/183749-how-to-copy-line-entry-in-xml-ccd-package

**Reply by jeffkirkley** [reply]
@pacmano1 Thanks for providing that.  I never heard anything more on the Mirth forums site from you, so I posted here.  Anything to add on how I can resolve getting rid of the xmlns entry?

**Reply by pacmano1** [reply]
I posted a working solution except for the fact your CDA is wrapped in extra tags.   I would remove those tags entirely.    

Since you went another direction, I deferred to others to help you as I will defer on your post here asking for a different soluton.

What is usually more helpful when you post is more detail.  e.g. "I am picking up an XML file from disk that has a CDA in it wrapped in extra tags".  In that case I would have suggested a preprocessor to strip those extra tags entirely, for example:

in the preprocessor:

```JS
var rawMessage = message;
rawMessage = rawMessage.replace(/<\s*xml\s*>/gi, '').replace(/<\s*\/\s*xml\s*>/gi, '');e
rawMessage = rawMessage.replace(/<\s*document\s*>/gi, '').replace(/<\s*\/\s*document\s*>/gi, '');
return rawMessage.trim();  // trim() to remove leading/trailing whitespaces if any
```
and again the code that already works:
```JS
var currentID = msg['recordTarget']['patientRole']['id'][0];
msg['recordTarget']['patientRole'].insertChildAfter(currentID, <id extension="AnotherExtension" root="AnotherRoot"/>);
```

**Reply by jeffkirkley** [reply]
@pacmano1 I appreciate your reply.  Again, begging your pardon, but I literally do not know the ins and outs with all of this and am just trying to get it ti work.  I understand what you said, but you are working with someone who has about 5 days of exposure to this system and yet I am trying to resolve a problem that should have been fixed yesterday.

I will take any and all coaching and assistance to try to get this to work in the proper way.  

Thank you again.

**Comment by tonygermano** [comment]
In the mirth data type properties for most xml based connector types, there is an option to "strip namespaces."

https://docs.nextgen.com/bundle/Mirth_User_Guide_42/page/connect/connect/topics/c_HL7_v3_x_Data_Type_connect_ug.html

**Reply by jeffkirkley** [reply]
I did try that, that did not provide good results.  It could be I need to do some pre or post formatting, but I do not that knowledge.  Again, for the record, this info I presented in my example is indicative of the type of data coming in on the wire from the distant end.

**Reply by tonygermano** [reply]
However, if you are adding values, then you probably do not want to do this, as the downstream system will likely expect the namespaces to still be present.

**Reply by jeffkirkley** [reply]
yes, and to my knowledge, the customer expectation is an XML file.

**Reply by tonygermano** [reply]
so, I don't think you want to be removing the namespaces. Are you just trying to make your existing solution more manageable? You can change the default namespace in mirth.

**Reply by jeffkirkley** [reply]
So, I do not believe I am able to change anything in Mirth.  From what I have seen so far, all work is done in the transformer scripts section for either the source or the destination.  For HL7 item, any of the prior-listed solutions would probably work.  It seems because this all XML (for CCD), that things are more trickier.

**Reply by tonygermano** [reply]
I see what your problem is. There are actually a couple with your example. So, the default xml namespace is `""`. When you create a new xml element that doesn't specify a namespace, it uses the default. So, `var addTemplateId = new XML('<id2/>');` is getting created with `xmlns=""`. If you don't want to change the default namespace, then you can do `var addTemplateId = <id2 xmlns="urn:hl7-org:v3"/>` instead of what you currently have.

Now, I couldn't use your code to add it to the document either, because it looks like you want id2 to be a child of id, but you have multiple `id` elements that match, so it can't add anything without you specifying which one you want to add it to.

**Reply by pacmano1** [reply]
I'd just reiterate that your example is not a CDA, it a CDA wrapped in other tags, and I would expect your customer to want only the part that begins with` <ClinicalDocument....>` and ends with `</ClinicalDocument>`

**Reply by jeffkirkley** [reply]
so, actually, the only way I could get it to work was as a child to <id>.  In all actuality, I would prefer what @pacmano1 provided initially:
But I could not do what he suggested: (below)


Assume the CDA is in tmp and the first line of the CDA starts with <ClinicalDocument...>

var currentID = tmp['recordTarget']['patientRole']['id'][0];
tmp['recordTarget']['patientRole'].insertChildAfter(currentID, <id extension="AnotherExtension" root="AnotherRoot"/>);​ 

Before:
![image](https://github.com/nextgenhealthcare/connect/assets/67973133/2ea10002-7a7d-484e-9f6e-dcccd098a457)

After:
![image](https://github.com/nextgenhealthcare/connect/assets/67973133/cb3a3be2-8a90-4a59-841d-9922d4557a0d)


I will try what you have just now suggested to see if it will make a difference.  

Question, will the output still have the "xmlns="?

**Reply by jeffkirkley** [reply]
Actually, it is an entire CCD record that contains a lot of stuff.  For the purposes of this public troubleshooting, I only included the part that is nested under ClinicalDocument and the <id> portion that needs to be modified.

**Reply by pacmano1** [reply]
If you are referring to my comment, I know your CDA is snippet of a real one.

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/38386c94-6721-4180-8dca-d56a579351f0)

**Reply by jeffkirkley** [reply]
so, the response from @tonygermano appeared to do the trick in removing the 'xmlns=""' entry.

THANK YOU!

Now, going back to what @pacmano1 suggested initially, is there anyway to get the transformed output to contain two <id> entries instead of the child <id2> I crafted?

**Reply by tonygermano** [reply]
So you want to add another `id` under `patientRole` instead of `id2` under `id`? This is the easiest way as long as you understand the caveats. You can also create your xml all in one step and fill in values with variables as shown below if you want.

```jsx
var idRoot =  '2.16.840.1.113883.10.20.22.1.1';
var addTemplateId = <id xmlns="urn:hl7-org:v3" extension="2023-08-15" root={idRoot} />;
// addTemplateId['@root'] = '2.16.840.1.113883.10.20.22.1.1';
// addTemplateId['@extension'] = '2023-08-15';

msg['document'].*::['ClinicalDocument'].*::['recordTarget'].*::['patientRole'].*::['id'] += addTemplateId;
```

This will:
1. take all of `id` elements that currently exist under patientRole
2. group them together
3. append addTemplateId to the end of the list
4. replace all of the `id`s in the original list with the new list in the place where the first `id` appeared.

**Reply by jeffkirkley** [reply]
So, I tried that, but am guessing I didn't get something right.  got this error on the destination transformer output:

DETAILS:              TypeError: Assignment to lists with more than one item is not supported

**Reply by jeffkirkley** [reply]
Scratch my last.  I had to remove the hard-coded entries below and just go with what you offered.  The output is now two separate <id> lines.

WOW!  THANK YOU!!!!!

I will continue to massage everything and report back here on the results.

**Reply by jeffkirkley** [reply]
So, one more tiny issue (hoping this can be easy to fix)......

For the sample output above (needed to assigningAuthority to the transformed <id> output:

<id extension="BLAH" root="2.16.840.1.113883.3.8294.4.141354861527044"/>
<id extension="AnotherExtension" root="AnotherRoot" assigningAuthority="XXXXX"/>


I have got the output looking correct, but the order of printed output is putting 'assigningAuthority' first in the list:

<id extension="BLAH" root="2.16.840.1.113883.3.8294.4.141354861527044"/>
<id assigningAuthority="XXXXX" extension="AnotherExtension" root="AnotherRoot" />

Is there a way to order this output for the new <id> to put 'assigningAuthority' last on the line?

**Reply by tonygermano** [reply]
I'm pretty sure it always alphabetizes the properties, so likely the answer to that question is no.

**Reply by tonygermano** [reply]
Property order should not matter like element order often does.

**Reply by jeffkirkley** [reply]
I gotcha.  I was thinking that was the case as well, but had to ask.  Thanks.
